### PR TITLE
Add start_index and end_index

### PR DESCRIPTION
**Pull request by manuscle**
So you can show : Item [start_index] to [end_index] on [total_items]
or if you prefer : Item 20 to 29 on 100

### CODE REVIEW

File: test/Plugin/HalTest.php [hunk 1]
```diff
@@ -1010,12 +1010,16 @@ public function testRenderingPaginatorCollectionRendersPaginationAttributes()
...
+            'start_index',
+            'end_index',
         ];
         $this->assertEquals($expected, array_keys($rendered));
         $this->assertEquals(100, $rendered['total_items']);
         $this->assertEquals(3, $rendered['page']);
         $this->assertEquals(10, $rendered['page_count']);
         $this->assertEquals(10, $rendered['page_size']);
+        $this->assertEquals(20, $rendered['start_index']);
+        $this->assertEquals(29, $rendered['end_index']);
```
Comment (weierophinney) [inline test/Plugin/HalTest.php]: While this correctly ensures that the newly introduced attributes are present, we also need tests covering what happens in terms of which items are in the collection when those attributes are present in the initial request.

Ideally, all of these, including the above assertions, should be in a separate tests to ensure that _existing_ behavior remains unchanged, while the _new_ behavior is also present.

I'd have tests demonstrating:

- what happens when both items are present in the request
- what happens when only the `start_index` is present in the request (for instance, how does this interact with pagination?)
- what happens when only the `end_index` is present in the request (is this valid? if so, how does it interact with pagination?)

File: src/Plugin/Hal.php [hunk 2]
```diff
@@ -605,6 +605,14 @@ public function renderCollection(Collection $halCollection)
             $payload['page'] = ($payload['page_count'] > 0)
                 ? $halCollection->getPage()
                 : 0;
+            $payload['start_index'] = $payload['page'] > 0
+                ? ($payload['page_size'] * ($payload['page'] - 1)?:1)
```
Comment (weierophinney) [inline src/Plugin/Hal.php]: Add spaces around the `?:` operator, please.

File: src/Plugin/Hal.php [hunk 3]
```diff
@@ -605,6 +605,14 @@ public function renderCollection(Collection $halCollection)
...
+                : 0;
+            $payload['end_index'] = $payload['page'] > 0
+                ? ($payload['page'] < $payload['page_count']
+                    ? (($payload['page_size'] * $payload['page']) - 1)
+                    : $payload['total_items'])
+                : 0;
```
Comment (weierophinney) [inline src/Plugin/Hal.php]: This is pretty convoluted and hard to follow. Maybe write a dedicated method for calculating the value? That way you can return early when specific conditions are met.

Comment (manuscle) [reply src/Plugin/Hal.php]: Do you think it's should be better to get values from Paginator instead of Hal plugin?

Comment (weierophinney) [reply src/Plugin/Hal.php]: IIRC, we use the values in the zf-hal configuration to seed the paginator. That said, the total number of items and number of pages is generally from the paginator, so getting those from that would likely be the best idea.

**Comment by Wilt** [comment]
Should such logic not be a front end implementation?
I mean you have all the ingredients to calculate these "start_index" and "end_index" values on the front end side. I consider this interface specific (front-end) and IMO it is not the responsibility of the server to provide these values.

**Comment by weierophinney** [comment]
This repository has been closed and moved to laminas-api-tools/api-tools-hal; a new issue has been opened at https://github.com/laminas-api-tools/api-tools-hal/issues/4.

**Comment by weierophinney** [comment]
This repository has been moved to laminas-api-tools/api-tools-hal. If you feel that this patch is still relevant, please re-open against that repository, and reference this issue. To re-open, we suggest the following workflow:
 - Squash all commits in your branch (`git rebase -i origin/{branch}`)
 - Make a note of all changed files (`git diff --name-only origin/{branch}...HEAD
 - Run the laminas/laminas-migration tool on the code.
 - Clone laminas-api-tools/api-tools-hal to another directory.
 - Copy the files from the second bullet point to the clone of laminas-api-tools/api-tools-hal.
 - In your clone of laminas-api-tools/api-tools-hal, commit the files, push to your fork, and open the new PR.
 We will be providing tooling via laminas/laminas-migration soon to help  automate the process.
